### PR TITLE
lowerUTF8/upperUTF8 allow lower/uppercase characters occupy different number of bytes

### PR DESCRIPTION
**Pull request by gengliqi**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/8484

Problem Summary:
See https://github.com/pingcap/tiflash/issues/8484.

### What is changed and how it works?
1. lowerUTF8/upperUTF8 allow lower/uppercase characters occupy different number of bytes
2. lowerUTF8/upperUTF8 respect the row boundaries, which means invalid sequences do not influence other rows. Inspired by https://github.com/ClickHouse/ClickHouse/pull/41286.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix the bug that lowerUTF8/upperUTF8 do not allow lower/uppercase characters occupy different number of bytes
```

### CODE REVIEW

**Review by windtalker — APPROVED** [review]
LGTM

File: dbms/src/Functions/FunctionsString.cpp [hunk 1]
```diff
@@ -566,29 +612,70 @@ TIFLASH_DECLARE_MULTITARGET_FUNCTION_TP(
...
+            }
+        }
+
+        if unlikely (is_diff_offsets)
+        {
+            Int64 diff = 0;
+            for (size_t i = 0; i < dst_offsets.size(); ++i)
+            {
+                diff += (Int64)dst_offsets[i] - (Int64)src_offsets[i];
+                dst_offsets[i] = src_offsets[i] + diff;
```
Comment (SeaRise) [inline dbms/src/Functions/FunctionsString.cpp:676]: Please add some comments here.

**Review by SeaRise — APPROVED** [review]
Others LGTM

**Pull request merged**